src/camera/Camera.h

```diff
@@ -0,0 +1,295 @@
+#pragma once
+
+#include <memory>
+#include <mutex>
+#include <string>
+#include <thread>
+
+#include <opencv2/core.hpp>
+#include <opencv2/videoio.hpp>
+
+#include "CameraParams.h"
+
+/**
+   @defgroup camera Camera Access
+   @brief This module provides an interface for defining and accessing the
+   different cameras attached to the rover.
+
+   Cameras should be defined in a YAML file with a format that will be
+   documented soon, and the location of this file should be able to be
+   optionally specified at runtime.
+ */
+
+/**
+   @namespace cam
+   @brief Namespace for camera access system.
+ */
+namespace cam
+{
+
+/**
+   @addtogroup camera
+   @{
+ */
+
+/**
+	@name Configuration File Keys
+
+	The following are constants for the keys used in the camera configuration
+	files. See @ref cameraconfig for more details.
+*/
+/**@{*/
+/**
+   Config file key for camera filename.
+ */
+const std::string KEY_FILENAME = "filename";
+/**
+   Config file key for camera id.
+ */
+const std::string KEY_CAMERA_ID = "camera_id";
+/**
+   Config file key for intrinsic parameters.
+ */
+const std::string KEY_INTRINSIC_PARAMS = "intrinsic_params";
+/**
+   Config file key for extrinsic parameters.
+ */
+const std::string KEY_EXTRINSIC_PARAMS = "extrinsic_params";
+/**
+   Config file key for calibration information.
+ */
+const std::string KEY_CALIB_INFO = "calib_info";
+/**
+   Config file key for camera name.
+ */
+const std::string KEY_NAME = "name";
+/**
+   Config file key for camera description.
+ */
+const std::string KEY_DESCRIPTION = "description";
+/**@}*/
+
+/**
+	@brief Represents a camera on the rover, from which a video feed can be
+	retrieved.
+
+	The Camera class provides an interface for opening a camera attached to the
+	rover, and retrieving frames from it. Additionally, a Camera object contains
+	other information associated with the camera, such as a description, its
+	location on the rover, and its intrinsic camera parameters.
+
+	A Camera object provides concurrent access to the camera; that is, the
+	video feed will be updated in a separate thread and a client may retrieve a
+	frame at any time. To avoid getting duplicate frames, there is a concept of
+	a "frame number"; it will be returned upon retrieving a frame, and may be
+	used to check if the current frame held by the Camera is different from the
+	last one the client retrieved. The frame number will be different if two
+	frames (within a reasonable time interval) are different.
+
+	Additionally, if copied, Camera objects share ownership over a camera; when
+	the last Camera object is destroyed, the underlying camera will be
+	closed. This means multiple Camera objects may exist at different places in
+	the code that refer to the same underlying camera and share access to it.
+
+	@warning Note that intrinsic and extrinsic parameters are optional for a
+	Camera object; such Cameras may not be suitable for some computer vision
```

Comment:
"such Cameras" meaning, cameras without one or both of those parameters?

Reply:
Yeah, that is confusing wording on second thought.

Reply:
Fixed in 759bbaa2e85f16fe33c7aff486c34636fcb857bd.